### ISSUE TITLE
I/O MMU vvtdEnabled breaks virtual machine resources on vSphere 6.5

### ISSUE THREAD

**Issue by sofixa** [hashicorp/terraform-provider-vsphere#1422]
### Terraform Version

1.0, 0.15, w/e

### vSphere Provider Version
2.0

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### Debug Output

```
Error: error reconfiguring virtual machine: error reconfiguring virtual machine: ServerFaultCode: 
Unexpected element tag "vvtdEnabled" seen

while parsing serialized DataObject of type vim.vm.FlagInfo
at line 2, column 477

while parsing property "flags" of static type VirtualMachineFlagInfo

while parsing serialized DataObject of type vim.vm.ConfigSpec
at line 2, column 146

while parsing call information for method ReconfigVM_Task
at line 2, column 66

while parsing SOAP body
 at line 2, column 60
 
 while parsing SOAP envelope
 at line 2, column 0
 
 while parsing HTTP request for method reconfigure
 on object of type vim.VirtualMachine
 at line 1, column 0
 
   on .terraform/modules/SV-WWW/main.tf line 1, in resource "vsphere_virtual_machine" "vm":
    1: resource "vsphere_virtual_machine" "vm" {
 
```

### Expected Behavior

The VM apply should work

### Actual Behavior

An error with `Unexpected element tag "vvtdEnabled" seen`. 

### Steps to Reproduce

1. Have a 6.5 vCenter
2. apply a basic VM to it

### References

The issue seems to be due to #1287 , more precisely this part:

https://github.com/hashicorp/terraform-provider-vsphere/blob/da0c688494d86dcc29d1cd6c202c7fa952b034db/vsphere/virtual_machine_config_structure.go#L373-L397

Basically it just gets the flags' values, which are false by default. However, vSphere <6.7 doesn't know them at all, and that results in an error.

Could we fix this ASAP please? I can do the code if needed, it's mostly an if.

ping @bjoernf73 @koikonom 

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by koikonom** [comment]
Hi @sofixa, sorry for this. I merged a fix and I'll cut 2.0.1 within the day.

**Comment by sofixa** [comment]
Great, thanks a lot!

**Comment by sofixa** [comment]
@koikonom hi, the fix didn't work sadly :( 

```
- Finding hashicorp/vault versions matching "~> 2.18"...
- Installing hashicorp/vsphere v2.0.1...
```

```
│ Error: error reconfiguring virtual machine: error reconfiguring virtual machine: ServerFaultCode: 
│ Unexpected element tag "vvtdEnabled" seen
│ 
│ while parsing serialized DataObject of type vim.vm.FlagInfo
│ at line 2, column 472
│ 
│ while parsing property "flags" of static type VirtualMachineFlagInfo
│ 
│ while parsing serialized DataObject of type vim.vm.ConfigSpec
│ at line 2, column 146
│ 
│ while parsing call information for method ReconfigVM_Task
│ at line 2, column 66
│ 
│ while parsing SOAP body
│ at line 2, column 60
│ 
│ while parsing SOAP envelope
│ at line 2, column 0
│ 
│ while parsing HTTP request for method reconfigure
│ on object of type vim.VirtualMachine
│ at line 1, column 0


```

Doesn't this part need an if as well?: 
https://github.com/hashicorp/terraform-provider-vsphere/blob/da0c688494d86dcc29d1cd6c202c7fa952b034db/vsphere/virtual_machine_config_structure.go#L389-L397

**Comment by koikonom** [comment]
Let me see if I can build a test lab with 6.5.

**Comment by dcpc007** [comment]
Hi, just uprate, and try, and no more error using a VSphere 6.0 with provider in 2.0.1 
with 2.0.0 i had the bug

**Comment by koikonom** [comment]
Thanks for the feedback @dcpc007 . 

I was wondering if you and @sofixa could post a copy of your `virtual_machine` resource block, so I can compare and see what could probably trigger the issue.

Unfortunately I wasn't able to get my hands on a 6.5 installation to test the code yet.

**Comment by clementbernet** [comment]
Hello @koikonom,

here is the `virtual_machine` resource block we use with @sofixa:


```
resource "vsphere_virtual_machine" "vm" {
  name             = var.name
  num_cpus         = var.cpu
  memory           = var.memory * 1024
  resource_pool_id = data.vsphere_compute_cluster.compute_cluster.resource_pool_id
  datastore_id     = data.vsphere_datastore.datastore[0].id
  host_system_id   = var.host_system_id
  folder           = var.folder
  annotation       = var.notes

  dynamic "disk" {
    for_each = var.disk_size
    content {
      label            = "${var.name}-${disk.key}.vmdk"
      size             = disk.value
      eagerly_scrub    = var.eagerly_scrub
      thin_provisioned = var.thin_provisioned
      unit_number      = disk.key
    }
  }

  clone {
    template_uuid = data.vsphere_content_library_item.image.id
  }

  lifecycle {
    ignore_changes = [clone[0].template_uuid, disk[0].key, custom_attributes]
  }

  cpu_hot_add_enabled         = true
  memory_hot_add_enabled      = true
  wait_for_guest_net_routable = false

  dynamic "network_interface" {
    for_each = data.vsphere_network.network
    content {
      network_id   = network_interface.value.id
      adapter_type = var.network_interface_type
      ovf_mapping  = var.ovf_map[network_interface.key]
    }
  }

  custom_attributes = {
    (data.vsphere_custom_attribute.synapps_project_id.id) = (var.synapps_project_id)
  }

  extra_config = {
    "guestinfo.vmname" = var.name
  }
}

```

**Comment by avignat** [comment]
Same here :
```
vsphere_virtual_machine.vm: Still creating... [1m0s elapsed]
╷
│ Error: error reconfiguring virtual machine: error reconfiguring virtual machine: ServerFaultCode:
│ Unexpected element tag "vvtdEnabled" seen
│
│ while parsing serialized DataObject of type vim.vm.FlagInfo
│ at line 2, column 469
│
│ while parsing property "flags" of static type VirtualMachineFlagInfo
│
│ while parsing serialized DataObject of type vim.vm.ConfigSpec
│ at line 2, column 144
│
│ while parsing call information for method ReconfigVM_Task
│ at line 2, column 66
│
│ while parsing SOAP body
│ at line 2, column 60
│
│ while parsing SOAP envelope
│ at line 2, column 0
│
│ while parsing HTTP request for method reconfigure
│ on object of type vim.VirtualMachine
│ at line 1, column 0
│
│   with vsphere_virtual_machine.vm,
│   on main.tf line 39, in resource "vsphere_virtual_machine" "vm":
│   39: resource "vsphere_virtual_machine" "vm" {
│
╵```

**Comment by koikonom** [comment]
OK @sofixa is probably right and the second snippet he pointed out in https://github.com/hashicorp/terraform-provider-vsphere/issues/1422#issuecomment-858444784 is triggered in the event of an update operation.

If the case is that the folks reporting that it works are creating new VMs and the people reporting issues are updating existing ones then this can be fixed. 

Before I do that would it be possible to provide a debug log of a failed operation? Set the environment variable `TF_LOG` to `DEBUG`, run the apply step and attach the output here.

Thank you all so much for your help!

**Comment by avignat** [comment]
My run with `TF_LOG=DEBUG` 
https://bin.mycozy.space/?50730a792e499c13#EWKX7UciDsDkcqcn5zhbgzdqkLKAgi6SNCEL5sg4hDF3

**Comment by pepemikrobi** [comment]
Just hit exactly the same symptoms, vCenter 6.5 / TF 1.0.0 / provider 2.0.1

**Comment by koikonom** [comment]
Hello,

I pushed some more changes in https://github.com/hashicorp/terraform-provider-vsphere/tree/vvtb_65_v2 . 

Can someone who's experiencing the issue on vsphere 6.5 build the provider and test if that addresses the issue? 

- Clone the repo: `git clone https://github.com/hashicorp/terraform-provider-vsphere`
- Checkout the branch: `git checkout vvtb_65_v2`
- Build the provider: `make build` . After this step the provider binary will show up in your $GOPATH/bin directory. 
- Add the following in your ~/.terraformrc:
```
provider_installation {
    dev_overrides {
        "hashicorp/vsphere" = "/Users/kyriakos/go/bin"
    }
    direct {}
}
```

Once these steps are done, run `terraform init` before doing the plan and apply steps.

**Comment by clementbernet** [comment]
Hello @koikonom,

Apparently the fix doesn't work.

terraform init :
(even if Terraform says it's using hashicorp/vsphere v1.26.0, we have a warning message telling us that provider development overrides are in effect, which is exactly what we want)

```

Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/aws from the dependency lock file
- Reusing previous version of hashicorp/vault from the dependency lock file
- Reusing previous version of hashicorp/vsphere from the dependency lock file
- Using previously-installed hashicorp/aws v3.45.0
- Using previously-installed hashicorp/vault v2.20.0
- Using previously-installed hashicorp/vsphere v1.26.0

╷
│ Warning: Provider development overrides are in effect
│ 
│ The following provider development overrides are set in the CLI
│ configuration:
│  - hashicorp/vsphere in /home/clement/go/bin
│ 
│ Skip terraform init when using provider development overrides. It is not
│ necessary and may error unexpectedly.
╵

Terraform has been successfully initialized!
```

terraform plan & apply :

```
vsphere_virtual_machine.vm: Creating...
vsphere_virtual_machine.vm: Still creating... [10s elapsed]
vsphere_virtual_machine.vm: Still creating... [20s elapsed]
vsphere_virtual_machine.vm: Still creating... [30s elapsed]
╷
│ Error: error reconfiguring virtual machine: error reconfiguring virtual machine: ServerFaultCode: 
│ Unexpected element tag "vvtdEnabled" seen
│ 
│ while parsing serialized DataObject of type vim.vm.FlagInfo
│ at line 2, column 437
│ 
│ while parsing property "flags" of static type VirtualMachineFlagInfo
│ 
│ while parsing serialized DataObject of type vim.vm.ConfigSpec
│ at line 2, column 146
│ 
│ while parsing call information for method ReconfigVM_Task
│ at line 2, column 66
│ 
│ while parsing SOAP body
│ at line 2, column 60
│ 
│ while parsing SOAP envelope
│ at line 2, column 0
│ 
│ while parsing HTTP request for method reconfigure
│ on object of type vim.VirtualMachine
│ at line 1, column 0
│ 
│   with vsphere_virtual_machine.vm,
│   on vm.tf line 1, in resource "vsphere_virtual_machine" "vm":
│    1: resource "vsphere_virtual_machine" "vm" {
│ 
╵
```

I double-checked if I was using your branch _vvtb_65_v2_ before building the provider.

**Comment by avignat** [comment]
Exact same thing for me as @clementbernet

**Comment by pkqsun** [comment]
Exact same issue with provider v2.0.1, the fix didn't work as expected sadly.

# terraform  -v
Terraform v0.14.8
+ provider terraform.io/hashicorp/vsphere v2.0.1
# govc about
Name:         VMware vCenter Server
Vendor:       VMware, Inc.
Version:      6.5.0
Build:        17994927
OS type:      linux-x64
API type:     VirtualCenter
API version:  6.5
Product ID:   vpx
UUID:         614d896f-51cd-4e86-a0a6

kindly ping @koikonom

**Comment by koikonom** [comment]
Hello @avignat and @clementbernet, thanks for your help so far. This is very weird, and the fact that I still can't get my hands on vsphere 6.5 is definitely a problem I am trying to address.

In the mean time I've pushed another commit to the vvtb_65_v2 branch that enables additional logging around the version check. Can you please build it, run it again with TF_LOG set to DEBUG and send me back the output? I am interested in lines that contain the string `VERSIONDEBUG` in them.

Thanks again.

**Comment by clementbernet** [comment]
Hello @koikonom,

here it is: https://gist.github.com/clementbernet/cc5bc93fd5494f7c0d077e5fd6721bc1

All occurrences of `VERSIONDEBUG`:
```
2021-06-18T14:19:18.534+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.534+0200
2021-06-18T14:19:18.534+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v build number is greater than other build number: timestamp=2021-06-18T14:19:18.534+0200
2021-06-18T14:19:18.534+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.534+0200
2021-06-18T14:19:18.534+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v build number is greater than other build number: timestamp=2021-06-18T14:19:18.534+0200
2021-06-18T14:19:18.535+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.535+0200
2021-06-18T14:19:18.535+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.535+0200
2021-06-18T14:19:18.657+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.657+0200
2021-06-18T14:19:18.657+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.657+0200
2021-06-18T14:19:18.681+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.681+0200
2021-06-18T14:19:18.681+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.681+0200
2021-06-18T14:19:18.767+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.767+0200
2021-06-18T14:19:18.767+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.767+0200
2021-06-18T14:19:18.843+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.843+0200
2021-06-18T14:19:18.843+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.843+0200
2021-06-18T14:19:18.928+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:5, Patch:0, Build:16613358} vs other: viapi.VSphereVersion{Product:"VMware vCenter Server", Major:6, Minor:0, Patch:0, Build:0}: timestamp=2021-06-18T14:19:18.928+0200
2021-06-18T14:19:18.928+0200 [INFO]  provider.terraform-provider-vsphere: 2021/06/18 14:19:18 [DEBUG] VERSIONDEBUG: v is newer than other, returning true: timestamp=2021-06-18T14:19:18.928+0200
```

**Comment by koikonom** [comment]
Yes, this is very helpful! 

The bug is in the logic, since your instance reports a build number greater than 0 then the code assumes you're good to go, which is not quite true as we can see.

**Comment by koikonom** [comment]
Update the branch just now. Please give it another go.

**Comment by avignat** [comment]
I had the same output as @clementbernet (https://bin.mycozy.space/?71b5f2322ede6a38#CJ37LdZXKqYWhSeryiRCaQ3JBDd4pKmGRWzDZWA1H3yD)
I updated to the latest commit something different is broken now : 
```
vsphere_virtual_machine.vm: Creating...
vsphere_virtual_machine.vm: Still creating... [10s elapsed]
vsphere_virtual_machine.vm: Still creating... [20s elapsed]
vsphere_virtual_machine.vm: Still creating... [30s elapsed]
vsphere_virtual_machine.vm: Still creating... [40s elapsed]
vsphere_virtual_machine.vm: Still creating... [50s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m0s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m10s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m20s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m30s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m40s elapsed]
╷
│ Error: error sending customization spec: Customization of the guest operating system 'debian10_64Guest' is not supported in this configuration. Microsoft Vista (TM) and Linux guests with Logical Volume Manager are supported only for recent ESX host and VMware Tools versions. Refer to vCenter documentation for supported configurations.
│
│   with vsphere_virtual_machine.vm,
│   on main.tf line 39, in resource "vsphere_virtual_machine" "vm":
│   39: resource "vsphere_virtual_machine" "vm" {
│
╵
````

**Comment by koikonom** [comment]
@avignat was this working with the previous version of the provider? Do you get the same issue if you set the guest os type as ubuntu instead of debian ?

**Comment by clementbernet** [comment]
Working for me, good job @koikonom :smiley: :

```
vsphere_virtual_machine.vm: Creating...
vsphere_virtual_machine.vm: Still creating... [10s elapsed]
vsphere_virtual_machine.vm: Still creating... [20s elapsed]
vsphere_virtual_machine.vm: Still creating... [30s elapsed]
vsphere_virtual_machine.vm: Still creating... [40s elapsed]
vsphere_virtual_machine.vm: Still creating... [50s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m0s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m10s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m20s elapsed]
vsphere_virtual_machine.vm: Creation complete after 1m23s

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

```
Note that I'm not using customization spec.

**Comment by koikonom** [comment]
@clementbernet good news, I'll push a clean PR on Monday. Thank you so much for your help!
